cpp/include/raft/neighbors/detail/cagra/utils.hpp

```diff
@@ -150,4 +152,93 @@ struct gen_index_msb_1_mask {
 };
 }  // namespace utils
 
+/**
+ * Utility to sync memory from a host_matrix_view to a device_matrix_view
+ *
+ * In certain situations (UVM/HMM/ATS) host memory might be directly accessible on the
+ * device, and no extra allocations need to be performed. This class checks
+ * if the host_matrix_view is already accessible on the device, and only creates device
+ * memory and copies over if necessary. In memory limited situations this is preferable
+ * to having both a host and device copy
+ */
+template <typename T, typename IdxT>
+class device_matrix_view_from_host {
+ public:
+  device_matrix_view_from_host(raft::resources const& res, host_matrix_view<T, IdxT> host_view)
```

Comment:
cc @wphicks this pattern seems a lot like the `mdbuffer` to me. The goal here is to make a `device_mdspan` when the pointer can be accessed from device or copy memory to device when it can't.